src/org/rascalmpl/interpreter/TraversalEvaluator.java

```diff
@@ -431,10 +445,20 @@ private IValue traverseADTOnce(IValue subject, CaseBlockList casesOrRules,
 			  // this indicates that a nested rewrite has applied to replace a valid child with an invalid child
 			  throw new UnexpectedType(cons.getConstructorType().getFieldTypes(), tf.tupleType(args), (AbstractAST) null);
 		  }
-		  
+
+		  if (memoize) {
+			if (rcons == null) {
+				throw new RuntimeException("null rcons?");
+			}
+			memoizedResults.__put(cons, new TraverseResult(tr.matched, rcons, true));
+		  }
 		  return rcons;
 		}
 		else {
+			if (memoize) {
```

Comment:
Maybe this repeated block is good for a simple `finally` block; this would remove at least one clone and guarantee that the amb is cached if `memoize` is true on all return paths.

Reply:
The arguments to memoize are slightly different, so a single finally block requires slightly more complex code in the rest of the body in order to setup variables for the arguments. I think that would defeat the purpose.